fix assert failed when timeout during call rate_ddir

zqs-Oppenauer opened the pull request:
fix assert failed when timeout during call rate_ddir

Adding DDIR_TIMEOUT in enum `fio_ddir`, and `rate_ddir` returns it when fio timeouts. `set_io_u_file` will directly break out of the loop, and `fill_io_u` won't be called again, which causes assert to fail in `rate_ddir`, because td->rwmix_ddir is DDIR_INVAL.

 Signed-off-by: QingSong Zhu <zhuqingsong.0909@bytedance.com>

axboe commented:
This needs a MUCH better description in the actual commit message. There's nothing in there, you left it all on the github pull request message which won't go into the tree.

axboe commented:
Caveat - haven't attempted to reproduce this, but can't we just do:

```
diff --git a/io_u.c b/io_u.c
index 07e5bac5ef52..27c4d55fa4b7 100644
--- a/io_u.c
+++ b/io_u.c
@@ -792,7 +792,7 @@ static enum fio_ddir get_rw_ddir(struct thread_data *td)
 		 * https://github.com/axboe/fio/issues/1501#issuecomment-1418327049
 		 */
 		td->rwmix_ddir = ddir;
-	} else
+	} else if (ddir != DDIR_INVAL)
 		td->rwmix_ddir = rate_ddir(td, ddir);
 	return td->rwmix_ddir;
 }
```

and be done with it?

axboe commented:
Or probably this is the better way, same kind of thing:

```
diff --git a/io_u.c b/io_u.c
index 07e5bac5ef52..ea016c90e0e2 100644
--- a/io_u.c
+++ b/io_u.c
@@ -792,7 +792,7 @@ static enum fio_ddir get_rw_ddir(struct thread_data *td)
 		 * https://github.com/axboe/fio/issues/1501#issuecomment-1418327049
 		 */
 		td->rwmix_ddir = ddir;
-	} else
+	} else if (ddir_rw(ddir))
 		td->rwmix_ddir = rate_ddir(td, ddir);
 	return td->rwmix_ddir;
 }
```

zqs-Oppenauer commented:
> This needs a MUCH better description in the actual commit message. There's nothing in there, you left it all on the github pull request message which won't go into the tree.
> 
> You also need the Signed-off-by in there, no in the git pull request.

Thanks very much for your 

> This needs a MUCH better description in the actual commit message. There's nothing in there, you left it all on the github pull request message which won't go into the tree.
> 
> You also need the Signed-off-by in there, no in the git pull request.

Thank you very much for your reminder. I will add  description and Signed-off-by in the commit message.

axboe commented:
I did ponder whether DDIR_TIMEOUT would be a solution, since it's much cleaner than hacking in odd return values and needing to check it twice. That provides a way to pass this information back inline, rather than out-of-band. I think that proposed patch looks MUCH better than the original one.

zqs-Oppenauer commented:
> I did ponder whether DDIR_TIMEOUT would be a solution, since it's much cleaner than hacking in odd return values and needing to check it twice. That provides a way to pass this information back inline, rather than out-of-band. I think that proposed patch looks MUCH better than the original one.

Thanks very much, I have already updated this commit according to the proposed patch and added the description and Signed-off-by in the commit message.

axboe commented on io_u.c:
```diff
@@ -1419,6 +1419,11 @@ static long set_io_u_file(struct thread_data *td, struct io_u *io_u)
 		td_io_close_file(td, f);
 		io_u->file = NULL;
+
+		if (io_u->ddir == DDIR_TIMEOUT) {
```
No braces necessary here.

axboe commented:
Looks much better, just one minor style issue. Outside of that, please line wrap commit messages at around 72 chars. Long lines are hard to read in git log. With those two things tweaked, this should be ready to go in.

axboe commented:
Once I triggered CI, it failed. The cases where we have a switch based on ddir, you need to also add DDIR_TIMEOUT to the DDIR_INVAL cases.

axboe commented:
Thanks!
